### PR TITLE
Issue202 spectro x2 d crystal

### PR DESCRIPTION
**Pull request by Didou09**
Intermediate PR to update devel with new features regarding 2D X-Ray spectrometer.

These new features are in tofu/geom/_core_optics.py, _comp_optics.py and _plot_optics.py.

They were added as extra files because they include a lot of new things (routines, classes) and because this way it does not affect the basic classes.
They are not unit-tested yet.

This is an update just to make available some functionalities and avoid a single huge PR at the end. But there is no user (except me) for these at this point, so it won't affect anyone.

### CODE REVIEW

File: tofu/geom/_core.py [hunk 1]
```diff
@@ -4141,6 +4141,11 @@ def _complete_dX12(self, dgeom):
                         pinhole = dgeom['D'][:, 0] + k[0]*u[:, 0]
                         dgeom['pinhole'] = pinhole
 
+            if np.any(np.isnan(dgeom['D'])):
```
Comment (Didou09) [inline tofu/geom/_core.py]: This change to _core.py is just adding a safeguard (making it more robust with respect to ill-defined input from users)

File: tofu/imas2tofu/_core.py [hunk 2]
```diff
@@ -1508,10 +1508,10 @@ def add_ids_base(self, occ=None, idd=None,
                      user=user, tokamak=tokamak, version=version,
                      ref=ref, isget=isget, get=get)
 
-    def add_ids_for_synthdiag(self, ids=None, occ=None, idd=None,
```
Comment (Didou09) [inline tofu/imas2tofu/_core.py]: Minor changes also to make it more robust

File: tofu/utils.py [hunk 3]
```diff
@@ -639,7 +647,7 @@ def _get_exception(q, ids, qtype='quantity'):
 
 
 def load_from_imas(shot=None, run=None, user=None, tokamak=None, version=None,
-                   ids=None, Name=None, out=None, tlim=None, config=None,
+                   ids=None, Name=None, returnas=None, tlim=None, config=None,
```
Comment (Didou09) [inline tofu/utils.py]: out keyword replaced by returnas for clarity (will do it gradually everytime I see an opportunity)

File: tofu/utils.py [hunk 4]
```diff
@@ -657,14 +665,16 @@ def load_from_imas(shot=None, run=None, user=None, tokamak=None, version=None,
...
+    c0 = returnas is None or returnas in lok
     if not c0:
-        msg = "Arg out must be in %s"%str(lok)
+        msg = "Arg returnas must be in %s"%str(lok)
         raise Exception(msg)
 
     # -------------------
     # Prepare ids
-    assert ids is None or type(ids) in [list,str]
+    if type(ids) not in [list, str]:
```
Comment (Didou09) [inline tofu/utils.py]: Minor change for robustness

File: tofu/utils.py [hunk 5]
```diff
@@ -1020,7 +1030,7 @@ def load_from_imas(shot=None, run=None, user=None, tokamak=None, version=None,
         dout = dout[shot[0]]['Data'][0]
     elif nshot == 1 and nPla == 1:
         dout = dout[shot[0]]['Plasma2D'][0]
-    return out
```
Comment (Didou09) [inline tofu/utils.py]: Was not returning the good variable...

File: tofu/utils.py [hunk 6]
```diff
@@ -320,7 +320,15 @@ def save(obj, path=None, name=None, sep=None, deep=False, mode='npz',
     # Get stripped dictionnary
     deep = 'dict' if deep else 'ref'
     if sep is None:
-        sep = _SEP
+        if mode == 'mat':
```
Comment (Didou09) [inline tofu/utils.py]: If a .mat file is being saved, the separator must be '_', otherwise a '.' would corrupt the mat file (minor change for robustness)

**Comment by lasofivec** [comment]
@Didou09 I'll let you see if you want to keep the branch or not. As I understood, the changes are not over

**Comment by lasofivec** [comment]
I just realized this PR introduced a *.npz file in the root directory. I will erase it in a new PR...

**Pull request merged**